teamcity: failed test: bank/zerosum-splits

The following tests appear to have failed on master (roachtest): acceptance/bank/zerosum-splits

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+bank/zerosum-splits).

[#1507340](https://teamcity.cockroachdb.com/viewLog.html?buildId=1507340):


```
acceptance/bank/zerosum-splits
--- FAIL: roachtest/acceptance/bank/zerosum-splits (901.150s)
test artifacts and logs in: artifacts/acceptance/bank/zerosum-splits/run_1
	bank.go:366,bank.go:512,acceptance.go:69,test_runner.go:689: pq: range r25:/Table/53{-/1/152} [(n1,s1):1VOTER_OUTGOING, (n2,s2):2, (n4,s4):3, (n3,s3):4VOTER_INCOMING, next=5, gen=8] had non-voter replicas: (n1,s1):1VOTER_OUTGOING,(n2,s2):2,(n4,s4):3,(n3,s3):4VOTER_INCOMING

test artifacts and logs in: artifacts/acceptance/bank/zerosum-splits/run_1
	bank.go:366,bank.go:512,acceptance.go:69,test_runner.go:689: pq: range r25:/Table/53{-/1/152} [(n1,s1):1VOTER_OUTGOING, (n2,s2):2, (n4,s4):3, (n3,s3):4VOTER_INCOMING, next=5, gen=8] had non-voter replicas: (n1,s1):1VOTER_OUTGOING,(n2,s2):2,(n4,s4):3,(n3,s3):4VOTER_INCOMING





```

Please assign, take a look and update the issue accordingly.

--- a/pkg/storage/merge_queue.go
+++ b/pkg/storage/merge_queue.go
@@ -275,30 +275,18 @@ func (mq *mergeQueue) process(
 	}
 
 	{
-		store, db := lhsRepl.store, lhsRepl.store.DB()
+		store := lhsRepl.store
 		// AdminMerge errors if there is a learner or joint config on either
 		// side and AdminRelocateRange removes any on the range it operates on.
 		// For the sake of obviousness, just fix this all upfront.
 		var err error
-		lhsDesc, err = maybeLeaveAtomicChangeReplicas(ctx, store, lhsDesc)
+		lhsDesc, err = removeNonFullVoters(ctx, store, lhsDesc)
 		if err != nil {
 			log.VEventf(ctx, 2, `%v`, err)
 			return err
 		}
 
-		lhsDesc, err = removeLearners(ctx, db, lhsDesc)
-		if err != nil {
-			log.VEventf(ctx, 2, `%v`, err)
-			return err
-		}
-
-		rhsDesc, err = maybeLeaveAtomicChangeReplicas(ctx, store, rhsDesc)
-		if err != nil {
-			log.VEventf(ctx, 2, `%v`, err)
-			return err
-		}
-
-		rhsDesc, err = removeLearners(ctx, db, rhsDesc)
+		rhsDesc, err = removeNonFullVoters(ctx, store, rhsDesc)
 		if err != nil {
 			log.VEventf(ctx, 2, `%v`, err)
 			return err

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -2012,14 +2012,9 @@ func (s *Store) AdminRelocateRange(
 		ctx = client.ChangeReplicasCanMixAddAndRemoveContext(ctx)
 	}
 
-	// Step 0: Remove all learners so we don't have to think about them. We could
-	// do something smarter here and try to promote them, but it doesn't seem
-	// worth the complexity right now. Revisit if this is an issue in practice.
-	//
-	// Note that we can't just add the learners to removeTargets. The below logic
-	// always does add then remove and if the learner was in the requested
-	// targets, we might try to add it before removing it.
-	newDesc, err := removeLearners(ctx, s.DB(), &rangeDesc)
+	// Step 0: Remove everything that's not a full voter so we don't have to think
+	// about them.
+	newDesc, err := removeNonFullVoters(ctx, s, &rangeDesc)
 	if err != nil {
 		log.Warning(ctx, err)
 		return err
@@ -2342,10 +2337,16 @@ func (s *Store) relocateOne(
 	return ops, transferTarget, nil
 }
 
-func removeLearners(
-	ctx context.Context, db *client.DB, desc *roachpb.RangeDescriptor,
+func removeNonFullVoters(
+	ctx context.Context, store *Store, desc *roachpb.RangeDescriptor,
 ) (*roachpb.RangeDescriptor, error) {
 	origDesc := desc
+
+	desc, err := maybeLeaveAtomicChangeReplicas(ctx, store, desc)
+	if err != nil {
+		return nil, err
+	}
+
 	learners := desc.Replicas().Learners()
 	if len(learners) == 0 {
 		return desc, nil
@@ -2363,7 +2364,7 @@ func removeLearners(
 	// https://github.com/cockroachdb/cockroach/pull/40268
 	for i := range chgs {
 		var err error
-		desc, err = db.AdminChangeReplicas(ctx, desc.StartKey, *desc, chgs[i:i+1])
+		desc, err = store.DB().AdminChangeReplicas(ctx, desc.StartKey, *desc, chgs[i:i+1])
 		if err != nil {
 			return nil, errors.Wrapf(err, `removing learners from %s`, origDesc)
 		}

--- a/pkg/storage/replica_learner_test.go
+++ b/pkg/storage/replica_learner_test.go
@@ -792,7 +792,7 @@ func TestLearnerAndJointConfigFollowerRead(t *testing.T) {
 	check()
 }
 
-func TestLearnerAdminRelocateRange(t *testing.T) {
+func TestLearnerOrJointConfigAdminRelocateRange(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
 	ctx := context.Background()
@@ -812,22 +812,39 @@ func TestLearnerAdminRelocateRange(t *testing.T) {
 		_ = tc.AddReplicasOrFatal(t, scratchStartKey, tc.Target(2))
 	})
 
+	check := func(targets []roachpb.ReplicationTarget) {
+		require.NoError(t, tc.Server(0).DB().AdminRelocateRange(ctx, scratchStartKey, targets))
+		desc := tc.LookupRangeOrFatal(t, scratchStartKey)
+		voters := desc.Replicas().Voters()
+		require.Len(t, voters, len(targets))
+		sort.Slice(voters, func(i, j int) bool { return voters[i].NodeID < voters[j].NodeID })
+		for i := range voters {
+			require.Equal(t, targets[i].NodeID, voters[i].NodeID, `%v`, voters)
+			require.Equal(t, targets[i].StoreID, voters[i].StoreID, `%v`, voters)
+		}
+		require.Empty(t, desc.Replicas().Learners())
+		require.Empty(t, desc.Replicas().Filter(predIncoming))
+		require.Empty(t, desc.Replicas().Filter(predOutgoing))
+	}
+
 	// Test AdminRelocateRange's treatment of learners by having one that it has
 	// to remove and one that should stay and become a voter.
 	//
 	// Before: 1 (voter), 2 (learner), 3 (learner)
 	// After: 1 (voter), 2 (voter), 4 (voter)
-	targets := []roachpb.ReplicationTarget{tc.Target(0), tc.Target(1), tc.Target(3)}
-	require.NoError(t, tc.Server(0).DB().AdminRelocateRange(ctx, scratchStartKey, targets))
-	desc := tc.LookupRangeOrFatal(t, scratchStartKey)
-	voters := desc.Replicas().Voters()
-	require.Len(t, voters, len(targets))
-	sort.Slice(voters, func(i, j int) bool { return voters[i].NodeID < voters[j].NodeID })
-	for i := range voters {
-		require.Equal(t, targets[i].NodeID, voters[i].NodeID, `%v`, voters)
-		require.Equal(t, targets[i].StoreID, voters[i].StoreID, `%v`, voters)
-	}
-	require.Empty(t, desc.Replicas().Learners())
+	check([]roachpb.ReplicationTarget{tc.Target(0), tc.Target(1), tc.Target(3)})
+
+	// AdminRelocateRange should leave joint configs before doing its thing.
+	//
+	// Before: 1 (voter), 2 (voter), 4 (outgoing)
+	// After: 1 (voter), 2 (voter), 3 (voter)
+	atomic.StoreInt64(&ltk.replicaAddStopAfterJointConfig, 1)
+	atomic.StoreInt64(&ltk.replicationAlwaysUseJointConfig, 1)
+	desc := tc.RemoveReplicasOrFatal(t, scratchStartKey, tc.Target(3))
+	require.True(t, desc.Replicas().InAtomicReplicationChange(), desc)
+	require.Len(t, desc.Replicas().Filter(predOutgoing), 1)
+	atomic.StoreInt64(&ltk.replicaAddStopAfterJointConfig, 0)
+	check([]roachpb.ReplicationTarget{tc.Target(0), tc.Target(1), tc.Target(2)})
 }
 
 func TestLearnerAndJointConfigAdminMerge(t *testing.T) {

